### PR TITLE
Add fromXContent method to GetResponse

### PR DESCRIPTION
**Pull request by javanna**
Moved field values `toXContent` logic to `GetField` (from `GetResult`), which outputs its own fields, and can also parse them now. Also added `fromXContent` to `GetResult` and `GetResponse`.

 The start object and end object for the get response output have been moved to `GetResult#toXContent`, from the corresponding rest action. This makes it possible to have `toXContent` and `fromXContent` completely symmetric, as parsing requires looping till an end object is found which is weird when the corresponding `toXContent` doesn't print that object out.

### CODE REVIEW

File: buildSrc/src/main/resources/checkstyle_suppressions.xml [hunk 1]
```diff
@@ -350,7 +350,6 @@
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]index[/\\]fielddata[/\\]plain[/\\]ParentChildIndexFieldData.java" checks="LineLength" />
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]index[/\\]fielddata[/\\]plain[/\\]SortedNumericDVIndexFieldData.java" checks="LineLength" />
   <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]index[/\\]fielddata[/\\]plain[/\\]SortedSetDVOrdinalsIndexFieldData.java" checks="LineLength" />
-  <suppress files="core[/\\]src[/\\]main[/\\]java[/\\]org[/\\]elasticsearch[/\\]index[/\\]get[/\\]GetResult.java" checks="LineLength" />
```
Comment (nik9000) [inline buildSrc/src/main/resources/checkstyle_suppressions.xml]: ❤️

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 2]
```diff
@@ -247,23 +245,71 @@ public XContentBuilder toXContentEmbedded(XContentBuilder builder, Params params
...
-            builder.field(Fields._TYPE, type);
-            builder.field(Fields._ID, id);
+            builder.field(_INDEX, index);
+            builder.field(_TYPE, type);
+            builder.field(_ID, id);
             if (version != -1) {
-                builder.field(Fields._VERSION, version);
+                builder.field(_VERSION, version);
             }
             toXContentEmbedded(builder, params);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: I really don't like those... I'm wondering if it would be better to have a marker interface `ToInnerXContent` or something

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 3]
```diff
@@ -247,23 +245,71 @@ public XContentBuilder toXContentEmbedded(XContentBuilder builder, Params params
...
+                } else if (FIELDS.equals(currentFieldName)) {
+                    while(parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        GetField getField = GetField.fromXContent(parser);
+                        fields.put(getField.getName(), getField);
+                    }
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "unsupported field: " + currentFieldName);
+                }
+            }
+        }
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: It is not an issue in `GetResult` but maybe we'll need to introduce some kind of post-parsing validation? Like if we parse incompatible values (found == true but version is still -1) we should raise an exception?

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        GetField getField = new GetField("field", Arrays.asList("value1", "value2"));
+        String output = toXContent(getField, XContentType.JSON, true).utf8ToString();
+        assertEquals("{\"field\":[\"value1\",\"value2\"]}", output);
+    }
+
+    public void testGetFieldEqualsAndHashcode() {
+        checkEqualsAndHashCode(randomGetField().v1(), GetResponseTests::copyGetField, GetResponseTests::mutateGetField);
+    }
+
+    public void testGetResultToAndFromXContent() throws Exception {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: I think we could have an abstract `XContentESTestCase` that contains a set of helpers and test methods, so that the following methods testGetResultToAndFromXContent/testGetResultEqualsAndHashcode etc could go in their own `GetResultTests` class.

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+                    break;
+                case 2:
+                    Short randomShort = randomShort();
+                    values.add(randomShort);
+                    expectedValues.add(randomShort.intValue());
+                    break;
+                case 3:
+                    Byte randomByte = randomByte();
+                    values.add(randomByte);
+                    expectedValues.add(randomByte.intValue());
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: randomByte.intValue() -> randomByte

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                    break;
+                case 1:
+                    int randomInt = randomInt();
+                    values.add(randomInt);
+                    expectedValues.add(randomInt);
+                    break;
+                case 2:
+                    Short randomShort = randomShort();
+                    values.add(randomShort);
+                    expectedValues.add(randomShort.intValue());
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: randomShort.intValue() -> randomShort

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+                    break;
+                case 7:
+                    String randomString = randomUnicodeOfLengthBetween(3, 10);
+                    values.add(randomString);
+                    expectedValues.add(randomString);
+                    break;
+                case 8:
+                    String randomText = randomUnicodeOfLengthBetween(3, 10);
+                    values.add(new Text(randomText));
+                    expectedValues.add(randomText);
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: Why no Text instance here?

**Review by tlrx — CHANGES_REQUESTED** [review]
Thanks @javanna! I left some comments but most of them are open questions. I don't consider them as blockers for this PR to be merged in but I'd like your opinion on them.

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 2 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: this object is reused in quite a few places where we call this `toXContentEmbedded` method to print out only part of it within a different ancestor. I am not sure a new interface would help this specific case, cause this inner method is not just about starting the ancestor externally, but also printing out only part of the object. it may help other cases though.

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 8]
```diff
@@ -247,23 +245,71 @@ public XContentBuilder toXContentEmbedded(XContentBuilder builder, Params params
...
             }
             toXContentEmbedded(builder, params);
         }
+        builder.endObject();
         return builder;
     }
 
+    public static GetResult fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.nextToken();
+        assert token == XContentParser.Token.START_OBJECT;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: leftover, I never know whether I want to use asserts or exceptions. I also don't want to end up validating that our own response are well formed, cause we should have tests for that, but I guess these few situations need to be checked. The question is exceptions or asserts?

File: core/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java [hunk 9]
```diff
@@ -42,12 +42,20 @@ public final RestResponse buildResponse(Response response) throws Exception {
     }
 
     public final RestResponse buildResponse(Response response, XContentBuilder builder) throws Exception {
-        builder.startObject();
+        if (wrapInObject()) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java]: I would love to get rid of those but the default here is starting an object externally, which is conceptually wrong. With this change there is one less of those situations. I expect that we clean this up as we go. objects should be started within toXContent as much as possible, otherwise the writing and parsing code will not be symmetric.

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 6 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: that wouldn't quite work. The list is typed `<Object>` if I add `randomShort` the `short` becomes `Short`, but I need an `Integer` in the map.

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 4 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: I agree in principle, I admit I am scared of starting with factoring things out and make things too generic and shoot ourselves in the foot. I'd need to look at how much copy pasting that would save us, which should be clearer as soon as we have added a few of these fromXContent methods. @cbuescher WDYT?

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+                    break;
```
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: I think we need to add a value type here that doesn't render to a simple json value but to an object, e.g. GeoPoint.

**Review by cbuescher — CHANGES_REQUESTED** [review]
I added a few comments, most are minor but I think we need to look into what happens when the GetField objects have an xContent representation that isn't a single value.

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 8 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: I'd prefer exceptions in this case

File: core/src/main/java/org/elasticsearch/index/get/GetField.java [hunk 11]
```diff
@@ -90,4 +95,58 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        }
+        String fieldName = parser.currentName();
+        List<Object> values = new ArrayList<>();
+        XContentParser.Token token = parser.nextToken();
+        if (token != XContentParser.Token.START_ARRAY) {
+            throw new ParsingException(parser.getTokenLocation(),
+                    "expected " + XContentParser.Token.START_ARRAY + " - found " + parser.currentToken());
+        }
+        while(parser.nextToken() != XContentParser.Token.END_ARRAY) {
+            values.add(parser.objectText());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: good point. I have tested geo points manually and we don't actually print anything out if any or those are requested via stored_fields, which makes me think that they are not supported in GetField, but I had exactly the same doubts. Printing things out is always easy as we have all the info (class name etc.) to decide what to do, when reading back it is much trickier as we only have the json data type information. We do end up not being able to parse back what we write. It may be better to change how we print things out... like simply printing out objects by calling their toString, but I am not sure if that's a good idea at the moment

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 8 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: sounds good to me

File: core/src/main/java/org/elasticsearch/index/get/GetField.java (hunk 11 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: I just tested geo points manually on 5.1 and it does print out an array if type is set to "stored" and requested via "stored_fields". I will do some further testing, I never used this part of the api tbh.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: who knows what I tested then, I will have another look.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: What' really strange is that stored `geo_point` fields get printed as a flat array, so if you store two point like:
```
PUT twitter/tweet/1
{
  "counter": 1,
  "location": [
    {
      "lat": 0.45,
      "lon": 0.33
    },
    {
      "lon": 1,
      "lat": 1.5
    }
  ]
}
```
I get back:
```
GET twitter/tweet/1?stored_fields=location

{
  "_index": "twitter",
  "_type": "tweet",
  "_id": "1",
  "_version": 4,
  "found": true,
  "fields": {
    "location": [
      "0.45, 0.33",
      "1.5, 1.0"
    ]
  }
}
```

So it looks like its printed as array of plain values. From reading  `ContentBuilder#Writers -> #value(GeoPoint p) -> #latlon()`  would have expected an array of objects, each containing a lat/lon parameters. It seems tricky to see how these values actually get stored in the index and then written to the response, so its also hard to parse them? I'd at least add the option of parsing maps here, then I think we are on the safe side.

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 4 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: At this point I would try to avoid too much hierarchy in the tests. In the query refactoring we ended up centralizing a lot of helpers in `AbstractQueryTestCase` but this ended up as a huge (>1000 loc) class and created a bunch of dependencies between the individual tests. To reuse freuqently needed helper methods I'd rather use utility classes with static methods where possible instead of creating abstract base classes too early. That said, we might need to revisit that decision at a later point.

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java (hunk 3 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: ++

File: core/src/main/java/org/elasticsearch/action/get/GetResponse.java [hunk 12]
```diff
@@ -169,6 +176,23 @@ public void writeTo(StreamOutput out) throws IOException {
...
 
     @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        GetResponse getFields = (GetResponse) o;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/get/GetResponse.java]: yea

File: core/src/main/java/org/elasticsearch/action/get/GetResponse.java [hunk 13]
```diff
@@ -156,6 +158,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return getResult.toXContent(builder, params);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/get/GetResponse.java]: also MultiGetResponse was using this, but I just removed the startObject and endObject from it.

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.common.xcontent.XContentBuilder.DEFAULT_DATE_PRINTER;
+import static org.elasticsearch.test.EqualsHashCodeTestUtils.checkEqualsAndHashCode;
+
+public class GetResponseTests extends ESTestCase {
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: I wondered too... I don't personally like having 3 classes with 2 test methods each that depend on each other. I also wonder whether it makes sense to test the GetResponse, as well as GetResult and GetField individually. Should we rather only test GetResponse?

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 15]
```diff
@@ -247,23 +245,71 @@ public XContentBuilder toXContentEmbedded(XContentBuilder builder, Params params
...
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
         if (!isExists()) {
-            builder.field(Fields._INDEX, index);
-            builder.field(Fields._TYPE, type);
-            builder.field(Fields._ID, id);
-            builder.field(Fields.FOUND, false);
+            builder.field(_INDEX, index);
+            builder.field(_TYPE, type);
+            builder.field(_ID, id);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: done

File: core/src/main/java/org/elasticsearch/index/get/GetField.java [hunk 16]
```diff
@@ -90,4 +95,58 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        for (Object value : values) {
+            builder.value(value);
+        }
+        builder.endArray();
+        return builder;
+    }
+
+    public static GetField fromXContent(XContentParser parser) throws IOException {
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new ParsingException(parser.getTokenLocation(),
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: will do once you pushed your PR, will reuse those methods that you are introducing

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 14 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: done

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 10 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: I think we are good, as I explained above.

**Comment by javanna** [comment]
I pushed some updates, this should be ready now.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java [hunk 17]
```diff
@@ -36,34 +35,6 @@ private XContentParserUtils() {
...
-
-        String current = parser.currentName() != null ? parser.currentName() : "<null>";
-        if (current.equals(fieldName) == false) {
-            String message = "Failed to parse object: expecting field with name [%s] but found [%s]";
-            throw new ParsingException(parser.getTokenLocation(), String.format(Locale.ROOT, message, fieldName, current));
-        }
-        return t;
-    }
-
-    /**
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java]: sorry @tlrx while using these methods I changed them quite a bit, let me know what you think. I noticed one of them was unused, so I removed it.

File: core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        try (XContentParser parser = xContentType.xContent().createParser(actual)) {
+            Map<String, Object> finalMap = parser.map();
+            try (XContentParser parser2 = xContentType.xContent().createParser(expected)) {
+                Map<String, Object> originalMap = parser2.map();
+                assertEquals(originalMap, finalMap);
+            }
+        }
+    }
+
+    //TODO move this to some utility class? It should be useful in other tests too, whenever we parse stored fields
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java]: there's a couple of TODOs like this, I am not even sure it is worth doing, the main issue being I don't where to put the class and what to call it :)

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        BytesReference finalBytes = toXContent(parsedGetResponse, xContentType, false);
+        assertSameOutput(originalBytes, finalBytes, xContentType);
+    }
+
+    public void testToXContent() throws IOException {
+        {
+            GetResponse getResponse = new GetResponse(new GetResult("index", "type", "id", 1, true, new BytesArray("{ \"field1\" : " +
+                    "\"value1\", \"field2\":\"value2\"}"), Collections.singletonMap("field1", new GetField("field1",
+                    Collections.singletonList("value1")))));
+            String output = toXContent(getResponse, XContentType.JSON, false).utf8ToString();
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: Strings.toString(ToXContent, false) ? I don't know..

File: core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+        GetField expected = new GetField(fieldName, tuple.v2());
+        return Tuple.tuple(input, expected);
+    }
+
+    //TODO move this to some utility class
+    /**
+     * Asserts that the provided {@link BytesReference}s hold the same content. The comparison is done between the map
+     * representation of the provided objects.
+     */
+    public static void assertSameOutput(BytesReference expected, BytesReference actual, XContentType xContentType) throws IOException {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java]: I think we can remove the TODO; rename the method to something like `assertEqualsAsMap`, and change it to

```
        try (
                XContentParser expectedParser = xContentType.xContent().createParser(expected);
                XContentParser actualParser = xContentType.xContent().createParser(actual);
        ) {
            assertEquals(expectedParser.map(), actualParser.map());
        }
```

File: core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+    //TODO move this to some utility class? It should be useful in other tests too, whenever we parse stored fields
+    /**
+     * Returns a tuple containing random stored field values and their corresponding expected values once printed out
+     * via {@link ToXContent#toXContent(XContentBuilder, ToXContent.Params)} and parsed back via {@link XContentParser#objectText()}.
+     * Generates values based on what can get printed out. Stored fields values are retrieved from lucene and converted via
+     * {@link org.elasticsearch.index.mapper.MappedFieldType#valueForDisplay(Object)} to either strings, numbers or booleans.
+     *
+     * @param xContentType the content type, used to determine what the expected values are for float numbers.
+     */
+    public static Tuple<List<Object>, List<Object>> randomStoredFieldValues(XContentType xContentType) {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java]: I think it can be private for now and the TODO removed... If really needed outside of this test, we could maybe create a `RandomObjects` utilities class in the test framework and add all this kind of method there.

File: core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+                    Byte randomByte = randomByte();
+                    originalValues.add(randomByte);
+                    expectedParsedValues.add(randomByte.intValue());
+                    break;
+                case 4:
+                    double randomDouble = randomDouble();
+                    originalValues.add(randomDouble);
+                    expectedParsedValues.add(randomDouble);
+                    break;
+                case 5:
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java]: :crying_cat_face:

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+        //test that we can parse what we print out
+        GetResult parsedGetResult;
+        try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+            parsedGetResult = GetResult.fromXContent(parser);
+            assertNull(parser.nextToken());
+        }
+        assertEquals(expectedGetResult, parsedGetResult);
+        //print the parsed object out and test that the output is the same as the original output
+        BytesReference finalBytes = toXContent(parsedGetResult, xContentType, false);
+        assertSameOutput(originalBytes, finalBytes, xContentType);
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: Can we add a comment line like:
```
// GetResult can contain stored fields of various types, including floats, where order cannot be guaranteed. 
// Since a byte-to-byte comparison cannot be done we use assertSameOutput.
```

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+            builder.startObject();
+            addFields(builder, 0);
+            builder.endObject();
+            return builder.bytes();
+        } catch(IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void addFields(XContentBuilder builder, int currentDepth) throws IOException {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: Can we add a comment on what the method does?

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+                        builder.value(randomFieldValue());
+                    }
+                }
+                builder.endArray();
+            }
+        }
+    }
+
+    private static Object randomFieldValue() {
+        switch(randomIntBetween(0, 3)) {
```
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: Can be changed to randomFrom(randomAsciiOfLength(5), randomIntBetween(0,3), etc);

**Review by tlrx — APPROVED** [review]
I left minor comments, otherwise it LGTM.

File: core/src/main/java/org/elasticsearch/index/get/GetResult.java [hunk 26]
```diff
@@ -196,15 +208,6 @@ public GetField field(String name) {
         return fields.values().iterator();
     }
 
-    static final class Fields {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetResult.java]: I don't know, I think I did it because I wanted to reuse these fields and in general we tend to be removing these. It's a harmless change, not that we have to make it everywhere. I think I will :)

File: core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java (hunk 21 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/get/GetFieldTests.java]: I already know it is needed outside of this test, other apis support stored_fields etc. I will do that

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java (hunk 23 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: to clarify, I don't think that this is valid only for this case. Ordering doesn't have to do with floats. It has to do with using maps to store stuff. And we use maps in a lot of places. After all keys ordering doesn't matter in json, byte per byte comparison are not going to fly. That is why I proposed to move it out to some utility class.

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java (hunk 25 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: it's more concise, but we end up generating many random values that we don't use. We keep only one. I think I prefer the old fashioned switch.

File: core/src/test/java/org/elasticsearch/index/get/GetResultTests.java (hunk 24 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/get/GetResultTests.java]: done

File: core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java (hunk 19 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/get/GetResponseTests.java]: yea it should work

File: core/src/main/java/org/elasticsearch/index/get/GetField.java [hunk 27]
```diff
@@ -90,4 +96,52 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        }
+        builder.endArray();
+        return builder;
+    }
+
+    public static GetField fromXContent(XContentParser parser) throws IOException {
+        ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser::getTokenLocation);
+        String fieldName = parser.currentName();
+        List<Object> values = new ArrayList<>();
+        XContentParser.Token token = parser.nextToken();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: good catch

File: core/src/main/java/org/elasticsearch/index/get/GetField.java [hunk 28]
```diff
@@ -90,4 +96,52 @@ public void writeTo(StreamOutput out) throws IOException {
...
+            builder.value(value);
+        }
+        builder.endArray();
+        return builder;
+    }
+
+    public static GetField fromXContent(XContentParser parser) throws IOException {
+        ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser::getTokenLocation);
+        String fieldName = parser.currentName();
+        List<Object> values = new ArrayList<>();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/get/GetField.java]: done

File: core/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java (hunk 9 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java]: see #16347 I will go back to this soon

**Review by cbuescher — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java [hunk 29]
```diff
@@ -349,4 +349,22 @@ public static void writeRawField(String field, BytesReference source, XContentBu
             builder.rawField(field, source);
         }
     }
+
+    /**
+     * Returns the bytes that represent the XContent output of the provided {@link ToXContent} object, using the provided
+     * {@link XContentType}. Wraps the output into a new anonymous object depending on the value of the wrapInObject argument.
+     */
+    public static BytesReference toXContent(ToXContent toXContent, XContentType xContentType, boolean wrapInObject) throws IOException {
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java]: Can we have a test for this?

**Review by tlrx — APPROVED** [review]

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java [hunk 30]
```diff
@@ -766,4 +770,62 @@ public static void assertDirectoryExists(Path dir) {
...
+    public static void assertEquivalent(BytesReference expected, BytesReference actual, XContentType xContentType) throws IOException {
+        //we tried comparing byte per byte, but that didn't fly for a couple of reasons:
+        //1) whenever anything goes through a map while parsing, ordering is not preserved, which is perfectly ok
+        //2) Jackson SMILE parser parses floats as double, which then get printed out as double (with double precision)
+        try (XContentParser actualParser = xContentType.xContent().createParser(actual)) {
+            Map<String, Object> actualMap = actualParser.map();
+            replaceBytesArrays(actualMap);
+            try (XContentParser expectedParser = xContentType.xContent().createParser(expected)) {
+                Map<String, Object> expectedMap = expectedParser.map();
+                replaceBytesArrays(expectedMap);
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: I removed this shenanigans on master: https://github.com/elastic/elasticsearch/commit/38914f17edff05801ccda1f818bf7fef0cd33896 cc @tlrx @cbuescher
